packages/ckeditor5-link/src/linkimageediting.js

```diff
@@ -106,23 +106,25 @@ export default class LinkImageEditing extends Plugin {
 // @returns {Function}
 function upcastLink( editor ) {
 	const isImageInlinePluginLoaded = editor.plugins.has( 'ImageInlineEditing' );
-	const linkUtils = editor.plugins.get( 'ImageUtils' );
+	const imageUtils = editor.plugins.get( 'ImageUtils' );
 
 	return dispatcher => {
 		dispatcher.on( 'element:a', ( evt, data, conversionApi ) => {
 			const viewLink = data.viewItem;
-			const imageInLink = getFirstImage( viewLink );
+			const imageInLink = imageUtils.findViewImgElement( viewLink );
 
 			if ( !imageInLink ) {
 				return;
 			}
 
-			const blockImageView = imageInLink.findAncestor( element => linkUtils.isBlockImageView( element ) );
+			const blockImageView = imageInLink.findAncestor( element => imageUtils.isBlockImageView( element ) );
 
-			// There are two possible cases to consider here
+			// There are four possible cases to consider here
 			//
 			// 1. A "root > ... > figure.image > a > img" structure.
-			// 2. A "root > ... > block > a > img" structure.
+			// 2. A "root > ... > figure.image > a > picture > img" structure.
+			// 3. A "root > ... > block > a > img" structure.
+			// 4. A "root > ... > block > a > picture > img" structure.
 			//
 			// but the latter should only be considered by this converter when the inline image plugin
```

Comment:
Nice catch in docs 😛 As for the "plain, linked...", yeah it still makes sense because `<picture>` is just a container for presentation shenanigans and does not change the fact this is just an inline image.